### PR TITLE
Profilers

### PR DESCRIPTION
**Pull request by AradCarmi**
## Related Issues
fixes: [link to the issue](https://jira-dc.paloaltonetworks.com/browse/CIAC-11474)

## Description
A new Profiler decorator to debug automation runs.

### CODE REVIEW

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py [hunk 1]
```diff
@@ -12270,8 +12283,127 @@ def get_server_config():
...
+            function_to_profile()
+
+    :param func: The function to be profiled.
+    :return: The profiled function.
+    """
+    import cProfile
+    import threading
+
+    def profiler_wrapper(*args, **kwargs):
+        """
```
Comment (JudahSchwartz) [inline Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:12305]: we should call to the demisto support_multithreading function in the beginning of this decorator

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py [hunk 2]
```diff
@@ -12270,8 +12283,127 @@ def get_server_config():
...
+            :param signal_event: The event to signal when profiling is complete.
+            :param results: A shared object to store the results of the profiled function.
+            :param args: The positional arguments to be passed to the profiled function.
+            :param kwargs: The keyword arguments
+            """
+            profiler.enable()
+            demisto.debug("Profiler started.")
+            try:
+                results["function_results"] = func(*args, **kwargs)
+            except Exception as e:
```
Comment (JudahSchwartz) [inline Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:12380]: do we need to explicitly catch exceptions thrown in the function? `try: finally` will be enough to make sure we are freeing the profiler and we dont need to explicitly deal with waht the function we are calling does

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py [hunk 3]
```diff
@@ -12270,8 +12283,127 @@ def get_server_config():
...
+                """
+                import tempfile
+                # Create a temporary file to store profiling data
+                with tempfile.NamedTemporaryFile(delete=False) as temp_file:
+                    profiler.dump_stats(temp_file.name)
+                    temp_file_path = temp_file.name
+
+                # Read the profiling data from the temporary file
+                with open(temp_file_path, 'rb') as f:
+                    profiling_results = f.read()
```
Comment (JudahSchwartz) [inline Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:12333]: double check we cant write to a StringIO or something as a file instead, so we dont need the overhead of writing to disk

**Review by JudahSchwartz — COMMENTED** [review]
looks great! A couple comments 🚀

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py (hunk 1 above)
Comment (AradCarmi) [reply Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:12305]: I'm not sure we need it as the integration itself doesn't run in multithreaded. The integration code itself runs on the secondary thread.

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py (hunk 2 above)
Comment (AradCarmi) [reply Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:12380]: Nice, didn't know we could do it! 🚀

Comment (AradCarmi) [reply Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:12380]: But the issue is that we don't raise the exception we got from the profiled function.

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py (hunk 3 above)
Comment (AradCarmi) [reply Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:12333]: The StringIO is creating the desired file but seems to be corrupted as I can't open it with the `snakeviz`.

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py [hunk 4]
```diff
@@ -12287,8 +12300,128 @@ def get_server_config():
...
+                demisto.results(fileResult('profiling_stats.prof', profiling_results))
+
+            # 5 minutes in nanoseconds
+            default_timeout = 60 * 5 * 1e9
+
+            # The system timeout configuration.
+            timeout_nanoseconds = demisto.callingContext["context"].get("TimeoutDuration") or default_timeout
+            timeout_seconds = timeout_nanoseconds / 1e9
+            failed_on_timeout = False
+            event_set = signal_event.wait(timeout_seconds - 5)
```
Comment (JudahSchwartz) [inline Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:12376]: this is waiting 5 minutes - 5 seconds? or - 5 milliseconds?

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py [hunk 5]
```diff
@@ -12287,8 +12300,128 @@ def get_server_config():
...
+        support_multithreading()
+        results = {}
+        profiler = cProfile.Profile()
+        signal_event = threading.Event()
+        profiler_thread = threading.Thread(target=function_runner, args=(func, profiler, signal_event, results) + args,
+                                           kwargs=kwargs)
+        profiler_thread.start()
+
+        failed_on_timeout = profiler_function(signal_event, profiler)
+        if failed_on_timeout:
```
Comment (JudahSchwartz) [inline Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:12412]: do we need an eception? I think the original error will propogate through this. Lets just log that this happened in demisto.error (and double check that this works)

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py [hunk 6]
```diff
@@ -12287,8 +12300,128 @@ def get_server_config():
...
+            raise DemistoException(f"The profiled function '{func.__name__}' failed due to a timeout.")
+        if results.get("function_results"):
+            res = results.get("function_results")
+            if isinstance(res, Exception):
+                raise res
+            return res
+        return None
```
Comment (JudahSchwartz) [inline Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:12419]: is this still needed?

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py (hunk 5 above)
Comment (AradCarmi) [reply Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:12412]: I will try to remove this and test it out.

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py (hunk 6 above)
Comment (AradCarmi) [reply Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:12419]: No, it will be removed.

**Comment by kobymeir** [comment]
### This PR was automatically updated by a [GitHub Action](https://github.com/demisto/content/actions/runs/12122003002)
- **Base** pack version was bumped to **1.36.2**.

To stop automatic version bumps, add the `ignore-auto-bump-version` label to the github PR.

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py [hunk 7]
```diff
@@ -12369,8 +12382,121 @@ def get_server_config():
     return server_config
 
 
-from DemistoClassApiModule import *     # type:ignore [no-redef]  # noqa:E402
+def content_profiler(func):
+    """
+    A decorator for profiling the execution time and performance of a function.
+
+    This decorator is useful for identifying performance bottlenecks and understanding the time complexity of your code.
```
Comment (JudahSchwartz) [inline Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:12389]: would add a line here about how to display the .prof file after downloading it

**Review by JudahSchwartz — APPROVED** [review]
Fantastic! Looks great.

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py (hunk 7 above)
Comment (JudahSchwartz) [reply Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:12389]: yeah just a line saying - file can be opened like `pip install snakeviz; snakveiz file.prof`

**Pull request merged**